### ISSUE TITLE
Out-of-sequence index in token array

### ISSUE THREAD

**Issue by neilnelson** [stanfordnlp/CoreNLP#1053]
The issue below is that the index value in the tokens array in the same sentence goes from 34 to 3 obtaining an ambiguity in using the index value since the same lower index values have already been used for earlier tokens in the sentence. The enhancedPlusPlusDependencies array uses the out-of-sequence index value which would be incorrect if the earlier tokens having the duplicate index value but the wrong token were selected by a program not aware of the ambiguity. If the index value increased for each next token as I have seen it in all other cases, this ambiguity would be avoided.

Sentence = 'We can count on the fingers of one of our hands all those worthy of poetic fame who now live in this great country of intellectual and civilized men,--one for every ten millions.'

Out of sequence index values occur at 'ten millions'.

In php the java-server call sequence = 'wget --post-data \''.$urlencode_paragraph.'\' \'localhost:9000/?properties={"annotators":"'.$annotators_txt.'","outputFormat":"json"}\' -O -'

$annotators_txt = 'tokenize,ssplit,pos,lemma,ner,regexner,parse,depparse,coref,dcoref,relation,kbp,natlog,udfeats,truecase,openie,quote'

Server run command = taskset -c 0-15 java -cp "/home/nnelson/Documents/downloads/CoreNLP/stanford-corenlp-full-2020-04-20/*" edu.stanford.nlp.pipeline.StanfordCoreNLPServer

End of tokens array
        [33]=>
        array(13) {
          ["index"]=>
          int(34)
          ["word"]=>
          string(5) "every"
...
        }
        [34]=>
        array(14) {
          ["index"]=>
          int(3)
          ["word"]=>
          string(3) "ten"
...
        }
        [35]=>
        array(14) {
          ["index"]=>
          int(4)
          ["word"]=>
          string(8) "millions"
...
        }
        [36]=>
        array(13) {
          ["index"]=>
          int(37)
          ["word"]=>
          string(1) "."
...
        }

In enhancedPlusPlusDependencies array
        [3]=>
        array(5) {
          ["dep"]=>
          string(6) "nummod"
          ["governor"]=>
          int(4)
          ["governorGloss"]=>
          string(8) "millions"
          ["dependent"]=>
          int(3)
          ["dependentGloss"]=>
          string(3) "ten"

**Comment by AngledLuffa** [comment]
I'm not able to reproduce this, either using CoreNLP directly or using the stanza client.  For example, here are a couple things I've tried.  If you can adjust either of these to reproduce the issue, I'll take a look.

```
cat foo.txt
We can count on the fingers of one of our hands all those worthy of poetic fame who now live in this great country of intellectual and civilized men,--one for every ten millions.

java edu.stanford.nlp.pipeline.StanfordCoreNLP -annotators "tokenize,ssplit,pos,parse,depparse" -file foo.txt -outputFormat json
```

```
from stanza.server import CoreNLPClient

text = "We can count on the fingers of one of our hands all those worthy of poetic fame who now live in this great country of intellectual and civilized men,--one for every ten millions."

# set up the client
print('---')
print('starting up Java Stanford CoreNLP Server...')

with CoreNLPClient(annotators=['tokenize','ssplit','pos','parse','depparse'],
                   output_format = "json",
                   classpath="$CLASSPATH",
                   be_quiet=False) as client:
    # submit the request to the server
    ann = client.annotate(text)
    print(ann)
```

**Comment by neilnelson** [comment]
John Bauer, You are quite correct. Using only the annotators "tokenize,ssplit,pos,parse,depparse" will obtain the correct tokens array result.

Using the annotators "tokenize,ssplit,pos,lemma,ner,regexner,parse,depparse,coref,dcoref,relation,kbp,natlog,udfeats,truecase,openie,quote" will obtain my result.

**Comment by AngledLuffa** [comment]
I'm still getting blocks that look like

```
 {'dep': 'case', 'governor': 36, 'governorGloss': 'millions', 'dependent': 33, 'dependentGloss': 'for'}, {'dep': 'det', 'governor': 36, 'governorGloss': 'millions', 'dependent': 34, 'dependentGloss': 'every'}, {'dep': 'nummod', 'governor': 36, 'governorGloss': 'millions', 'dependent': 35, 'dependentGloss': 'ten'}, {'dep': 'nmod:for', 'governor': 32, 'governorGloss': 'one', 'dependent': 36, 'dependentGloss': 'millions'}
```

so I'm not sure what I should be looking for or how to make this happen.

**Comment by neilnelson** [comment]
I am using the Stanford Parser via the server in a terminal in Xubuntu Linux as
taskset -c 0-15 java -cp "/home/nnelson/Documents/downloads/CoreNLP/stanford-corenlp-full-2020-04-20/*" edu.stanford.nlp.pipeline.StanfordCoreNLPServer

Then in PHP I extract a sufficiently small portion of text from a Gutenberg book and feed it to the parser using wget. For this case the wget command is
wget --post-data 'We%20can%20count%20on%20the%20fingers%20of%20one%20of%20our%20hands%20all%20those%20worthy%20of%20poetic%20fame%20who%20now%20live%20in%20this%20great%20country%20of%20intellectual%20and%20civilized%20men%2C--one%20for%20every%20ten%20millions.' 'localhost:9000/?properties={"annotators":"tokenize,ssplit,pos,lemma,ner,regexner,parse,depparse,coref,dcoref,relation,kbp,natlog,udfeats,truecase,openie,quote
","outputFormat":"json"}' -O -

Note that the %20 and similar characters are from urlencoding.

The server processes the text and returns json text to the terminal as below when I run the above wget as a terminal command. Note, I am using the enhancedDependencies array, but I am not seeing a difference in the other two dependency arrays basicDependencies and enhancedDependencies.

      "enhancedPlusPlusDependencies": [
        {
          "dep": "ROOT",
          "governor": 0,
          "governorGloss": "ROOT",
          "dependent": 3,
          "dependentGloss": "count"
        },
        {
          "dep": "nsubj",
          "governor": 3,
          "governorGloss": "count",
          "dependent": 1,
          "dependentGloss": "We"
        },
        {
          "dep": "aux",
          "governor": 3,
          "governorGloss": "count",
          "dependent": 2,
          "dependentGloss": "can"
        },
        {
          "dep": "nummod",
          "governor": 4,
          "governorGloss": "millions",
          "dependent": 3,
          "dependentGloss": "ten"
        },
        {
          "dep": "case",
          "governor": 6,
          "governorGloss": "fingers",
          "dependent": 4,
          "dependentGloss": "on"
        },
        {
          "dep": "nmod:for",
          "governor": 32,
          "governorGloss": "one",
          "dependent": 4,
          "dependentGloss": "millions"
        },
        {
          "dep": "det",
          "governor": 6,
          "governorGloss": "fingers",
          "dependent": 5,
          "dependentGloss": "the"
        },
        {
          "dep": "obl:on",
          "governor": 3,
          "governorGloss": "count",
          "dependent": 6,
          "dependentGloss": "fingers"
        },
        {
          "dep": "case",
          "governor": 8,
          "governorGloss": "one",
          "dependent": 7,
          "dependentGloss": "of"
        },
        {
          "dep": "nmod:of",
          "governor": 6,
          "governorGloss": "fingers",
          "dependent": 8,
          "dependentGloss": "one"
        },
        {
          "dep": "case",
          "governor": 11,
          "governorGloss": "hands",
          "dependent": 9,
          "dependentGloss": "of"
        },
        {
          "dep": "nmod:poss",
          "governor": 11,
          "governorGloss": "hands",
          "dependent": 10,
          "dependentGloss": "our"
        },
        {
          "dep": "nmod:of",
          "governor": 8,
          "governorGloss": "one",
          "dependent": 11,
          "dependentGloss": "hands"
        },
        {
          "dep": "dep",
          "governor": 8,
          "governorGloss": "one",
          "dependent": 12,
          "dependentGloss": "all"
        },
        {
          "dep": "dep",
          "governor": 12,
          "governorGloss": "all",
          "dependent": 13,
          "dependentGloss": "those"
        },
        {
          "dep": "amod",
          "governor": 13,
          "governorGloss": "those",
          "dependent": 14,
          "dependentGloss": "worthy"
        },
        {
          "dep": "case",
          "governor": 17,
          "governorGloss": "fame",
          "dependent": 15,
          "dependentGloss": "of"
        },
        {
          "dep": "amod",
          "governor": 17,
          "governorGloss": "fame",
          "dependent": 16,
          "dependentGloss": "poetic"
        },
        {
          "dep": "obl:of",
          "governor": 14,
          "governorGloss": "worthy",
          "dependent": 17,
          "dependentGloss": "fame"
        },
        {
          "dep": "obj",
          "governor": 20,
          "governorGloss": "live",
          "dependent": 17,
          "dependentGloss": "fame"
        },
        {
          "dep": "ref",
          "governor": 17,
          "governorGloss": "fame",
          "dependent": 18,
          "dependentGloss": "who"
        },
        {
          "dep": "advmod",
          "governor": 20,
          "governorGloss": "live",
          "dependent": 19,
          "dependentGloss": "now"
        },
        {
          "dep": "acl:relcl",
          "governor": 17,
          "governorGloss": "fame",
          "dependent": 20,
          "dependentGloss": "live"
        },
        {
          "dep": "case",
          "governor": 24,
          "governorGloss": "country",
          "dependent": 21,
          "dependentGloss": "in"
        },
        {
          "dep": "det",
          "governor": 24,
          "governorGloss": "country",
          "dependent": 22,
          "dependentGloss": "this"
        },
        {
          "dep": "amod",
          "governor": 24,
          "governorGloss": "country",
          "dependent": 23,
          "dependentGloss": "great"
        },
        {
          "dep": "obl:in",
          "governor": 20,
          "governorGloss": "live",
          "dependent": 24,
          "dependentGloss": "country"
        },
        {
          "dep": "case",
          "governor": 29,
          "governorGloss": "men",
          "dependent": 25,
          "dependentGloss": "of"
        },
        {
          "dep": "amod",
          "governor": 29,
          "governorGloss": "men",
          "dependent": 26,
          "dependentGloss": "intellectual"
        },
        {
          "dep": "cc",
          "governor": 28,
          "governorGloss": "civilized",
          "dependent": 27,
          "dependentGloss": "and"
        },
        {
          "dep": "conj:and",
          "governor": 26,
          "governorGloss": "intellectual",
          "dependent": 28,
          "dependentGloss": "civilized"
        },
        {
          "dep": "amod",
          "governor": 29,
          "governorGloss": "men",
          "dependent": 28,
          "dependentGloss": "civilized"
        },
        {
          "dep": "nmod:of",
          "governor": 24,
          "governorGloss": "country",
          "dependent": 29,
          "dependentGloss": "men"
        },
        {
          "dep": "punct",
          "governor": 13,
          "governorGloss": "those",
          "dependent": 30,
          "dependentGloss": ","
        },
        {
          "dep": "punct",
          "governor": 13,
          "governorGloss": "those",
          "dependent": 31,
          "dependentGloss": "--"
        },
        {
          "dep": "dep",
          "governor": 13,
          "governorGloss": "those",
          "dependent": 32,
          "dependentGloss": "one"
        },
        {
          "dep": "case",
          "governor": 4,
          "governorGloss": "millions",
          "dependent": 33,
          "dependentGloss": "for"
        },
        {
          "dep": "det",
          "governor": 4,
          "governorGloss": "millions",
          "dependent": 34,
          "dependentGloss": "every"
        },
        {
          "dep": "punct",
          "governor": 3,
          "governorGloss": "count",
          "dependent": 37,
          "dependentGloss": "."
        }
      ],

It looks like we need to sync what we are doing. For example, if you are on Linux there is a good chance you could start the parser in one terminal window as follows replacing the path (??) as needed. 
java -cp "/??/CoreNLP/stanford-corenlp-full-2020-04-20/*" edu.stanford.nlp.pipeline.StanfordCoreNLPServer

You then may be able to submit the wget line from above in another terminal window and compare that output for the enhancedDependencies array given above. If your output was different than mine you could post the commands and output as I have done.

I am using java --version
openjdk 11.0.7 2020-04-14
OpenJDK Runtime Environment (build 11.0.7+10-post-Ubuntu-3ubuntu1)
OpenJDK 64-Bit Server VM (build 11.0.7+10-post-Ubuntu-3ubuntu1, mixed mode, sharing)

**Comment by AngledLuffa** [comment]
Alright, at this point I am able to reproduce the difference.  It seems to be in the relation annotator, maybe?

These two are different:

```
wget --post-data 'We%20can%20count%20on%20the%20fingers%20of%20one%20of%20our%20hands%20all%20those%20worthy%20of%20poetic%20fame%20who%20now%20live%20in%20this%20great%20country%20of%20intellectual%20and%20civilized%20men%2C--one%20for%20every%20ten%20millions.' 'localhost:9000/?properties={"annotators":"tokenize,ssplit,pos,lemma,ner,regexner,parse,depparse,coref,dcoref,relation","outputFormat":"json"}' -O foo.json
```

```
wget --post-data 'We%20can%20count%20on%20the%20fingers%20of%20one%20of%20our%20hands%20all%20those%20worthy%20of%20poetic%20fame%20who%20now%20live%20in%20this%20great%20country%20of%20intellectual%20and%20civilized%20men%2C--one%20for%20every%20ten%20millions.' 'localhost:9000/?properties={"annotators":"tokenize,ssplit,pos,lemma,ner,regexner,parse,depparse,coref,dcoref","outputFormat":"json"}' -O bar.json
```

**Comment by neilnelson** [comment]
Very good. I confirm that bar.json has the corrected index numbers for _ten_ and _millions_, while foo.json has the out-of-sequence index numbers. And as you indicated, the _relation_ annotator is the only difference.

**Comment by neilnelson** [comment]
It could be useful to have a number of sentences that exhibit this issue to help the debug process. Here are five more sentences I have confirmed that show the issue. At the moment I have 269 sentences for this issue, extracted by program, not manually confirmed, and can zip and post them to my github account for you to copy. Let me know.

It was the age of bad popes and quarrelsome nobles, and lazy monks and haughty bishops, and ignorant people, steeped in gloomy superstitions, two hundred years before America was discovered, and two hundred and fifty years before Michael Angelo erected the dome of St. Peter's.

His ancestors, boasting of Roman descent, belonged to the patriotic party,--the Guelphs, who had the ascendency in his early years,--that party which defended the claims of the Popes against the Emperors of Germany.

I allude to this notion of purgatory, which early entered into the creeds of theologians, and which was adopted by the Catholic Church, to show how powerful it was when human consciousness sought a relief from the pains of endless physical torments.

They are such as interested Thomas Aquinas and Bonaventura, Anselm and Bernard.

His last retreat was at Ravenna, and his last days were soothed with gentle attentions from Guido da Polenta, that kind duke who revived his fainting hopes.

**Comment by AngledLuffa** [comment]
Thanks.  I believe this should be sufficient to find the issue.  If not, I'll definitely ping here again

**Comment by AngledLuffa** [comment]
To summarize what is happening: the relation annotator copies the annotation and manipulates the copy in various ways, including changing the dependency graphs.  It's using a pointer rather than an actual copy for the dependencies, though.  This results in the dependencies for the original annotation changing.  

As far as I can tell, all of this is deliberate.  This was written 10 years ago, so the reasons for it happening are probably lost in time.  If I were to change it so it deep copied the trees & graphs making up the dependencies, this may very well have some affect later on in the pipeline which I don't know about.

In other words, I'm not quite sure what to do about it at this point...

**Comment by neilnelson** [comment]
Thank you for looking into this. I am now writing a correction for the parser data extracted to a MySql database for this issue, and on this end that data correction looks like a minor fix.

At least you have this issue in your records. You are welcome to close this issue according to your preference.

**Comment by AngledLuffa** [comment]
It's not resolved as far as I'm concerned.  I don't think the relation annotator should change the underlying dependencies, but I don't fully understand the consequences of changing that behavior.  Thanks for reporting this.

**Comment by AngledLuffa** [comment]
Frustratingly, I am no longer able to recreate this error.  Would you confirm that the following triggers it for you?

``` wget --post-data 'They%20are%20such%20as%20interested%20Thomas%20Aquinas%20and%20Bonaventura,%20Anselm%20and%20Bernard%.' 'localhost:9000/?properties={"annotators":"tokenize,ssplit,pos,lemma,ner,regexner,parse,depparse,coref,dcoref","outputFormat":"json"}' -O bar.json ```

``` wget --post-data 'They%20are%20such%20as%20interested%20Thomas%20Aquinas%20and%20Bonaventura,%20Anselm%20and%20Bernard%.' 'localhost:9000/?properties={"annotators":"tokenize,ssplit,pos,lemma,ner,regexner,parse,depparse,coref,dcoref,relation,kbp,natlog,udfeats,truecase,openie,quote","outputFormat":"json"}' -O foo.json```

**Comment by neilnelson** [comment]
In foo.json and not in bar.json in the tokens array. Issue bolded.

         {
           "index": 5,
           "word": "interested",
           "originalText": "interested",
           "lemma": "interested",
           "characterOffsetBegin": 17,
           "characterOffsetEnd": 27,
           "pos": "JJ",
           "ner": "O",
           "speaker": "PER0",
           "truecase": "LOWER",
           "truecaseText": "interested",
           "before": " ",
           "after": " "
         },
         {
*"index": 3,*
           "word": "Thomas",
           "originalText": "Thomas",
           "lemma": "Thomas",
           "characterOffsetBegin": 28,
           "characterOffsetEnd": 34,
           "pos": "NNP",
           "ner": "PERSON",
           "speaker": "PER0",
           "truecase": "INIT_UPPER",
           "truecaseText": "Thomas",
           "before": " ",
           "after": " "
         },
         {
*"index": 4,*
           "word": "Aquinas",
           "originalText": "Aquinas",
           "lemma": "Aquinas",
           "characterOffsetBegin": 35,
           "characterOffsetEnd": 42,
           "pos": "NNP",
           "ner": "PERSON",
           "speaker": "PER0",
           "truecase": "INIT_UPPER",
           "truecaseText": "Aquinas",
           "before": " ",
           "after": " "
         },
         {
           "index": 8,
           "word": "and",
           "originalText": "and",
           "lemma": "and",
           "characterOffsetBegin": 43,
           "characterOffsetEnd": 46,
           "pos": "CC",
           "ner": "O",
           "speaker": "PER0",
           "truecase": "LOWER",
           "truecaseText": "and",
           "before": " ",
           "after": " "
         },

Neil Nelson

**Comment by AngledLuffa** [comment]
It's really bizarre, but although I once reproduced this bug, it no longer is happening for me, even with a fresh download.

Would you try a candidate for our next release and see if you still get it to happen?

http://nlp.stanford.edu/software/stanford-corenlp-full-2020-07-09.zip

**Comment by neilnelson** [comment]
Not seeing the sequence issue for this 2020-07-09 version for this instance. I will run my larger set of sentences and see if they are corrected also.

**Comment by AngledLuffa** [comment]
The weird thing is, I don't think we changed anything.  Maybe one of the underlying models changed in a way that it hasn't been triggered yet.  If you find another sentence which makes it happen, please let us know.

**Comment by neilnelson** [comment]
Processed 3,335 sentences. Did not find any sentences with this issue.

**Comment by AngledLuffa** [comment]
Must have been a cosmic ray...
